### PR TITLE
dont break when there is an invalid share

### PR DESCRIPTION
**Pull request by icewind1991**
Instead just dont show the invalid share

cc @rullzer @LukasReschke @PVince81

### CODE REVIEW

**Comment by danimo** [comment]
Works here

File: apps/files_sharing/lib/sharedstorage.php [hunk 1]
```diff
@@ -78,15 +85,19 @@ private function init() {
...
-		$sourcePath = $this->ownerView->getPath($this->share['file_source']);
-		list($this->sourceStorage, $sourceInternalPath) = $this->ownerView->resolvePath($sourcePath);
-		$this->sourceRootInfo = $this->sourceStorage->getCache()->get($sourceInternalPath);
+		try {
+			Filesystem::initMountPoints($this->share['uid_owner']);
+			$sourcePath = $this->ownerView->getPath($this->share['file_source']);
+			list($this->sourceStorage, $sourceInternalPath) = $this->ownerView->resolvePath($sourcePath);
+			$this->sourceRootInfo = $this->sourceStorage->getCache()->get($sourceInternalPath);
+		} catch (\Exception $e) {
+			$this->logger->logException($e);
```
Comment (PVince81) [inline apps/files_sharing/lib/sharedstorage.php]: The trouble with this is that the code will continue running and will continue using the unitialized SharedStorage, so calling other methods will either fail silently or run into other issues when accessing uninitialized members.

Do we know the root cause of invalid shares reaching this part of the code ?

Comment (PVince81) [inline apps/files_sharing/lib/sharedstorage.php]: Is this cause by other buggy parts of the code ? If we consider that this situation is never supposed to happen except under buggy conditions, then I'm fine with this.

Else we might want to try and catch this a bit higher.

Comment (PVince81) [inline apps/files_sharing/lib/sharedstorage.php]: ... in places where we can explicitly tell the code to skip this SharedStorage altogether

Comment (PVince81) [inline apps/files_sharing/lib/sharedstorage.php]: ... or throw StorageNotAvailableException ? This one should be handled properly ?

Comment (icewind1991) [inline apps/files_sharing/lib/sharedstorage.php]: The logic added to `getCache` and `isValid` should properly block that

Comment (PVince81) [inline apps/files_sharing/lib/sharedstorage.php]: Ok thanks, makes sense

**Comment by PVince81** [comment]
Code makes sense.

> what do we do about invalid shares?

We need to find out why they exist in the first place. There's probably another bug somewhere that produces them. If they are still legitimate oprhaned/outdated shares, then we'll need a background job to clean them up eventually. There is already one to clear orphaned shares, might need to be extended to cover whichever case that is.

File: apps/files_external/lib/failedcache.php [hunk 2]
```diff
@@ -29,6 +29,17 @@
  * Storage placeholder to represent a missing precondition, storage unavailable
  */
 class FailedCache implements ICache {
+	private $visible;
```
Comment (LukasReschke) [inline apps/files_external/lib/failedcache.php]: PHPDoc would be neat here as well `/** @var bool $visible */`

File: apps/files_sharing/lib/sharedstorage.php [hunk 3]
```diff
@@ -31,6 +31,7 @@
 namespace OC\Files\Storage;
 
 use OC\Files\Filesystem;
+use OCA\Files_External\Lib\FailedCache;
```
Comment (LukasReschke) [inline apps/files_sharing/lib/sharedstorage.php]: Inter-App Dependency :see_no_evil:

**Comment by LukasReschke** [comment]
Patch also verified by a forum user at https://forum.owncloud.org/viewtopic.php?f=38&t=34198&p=109977#p109977:

> Hi AnybodyElse,
>
> Thank you for the quick & efficient support !
I applied the two patches and it's now solved.
>
>Rgds,
>a.
>
>PS: Is there a way to remove this missing folder ? It's indeed throwing errors at each page refresh :
> ```
{"reqId":"GCBmO3KOyu1RNpLNg8oL","remoteAddr":"2a02:a03f:2852:2500:857c:a818:a1b3:737f","app":"no app in context","message":"Exception: {\"Exception\":\"OCP\\\\Files\\\\NotFoundException\",\"Message\":\"File with id \\\"33084\\\" has not been found.\",\"Code\":0,\"Trace\":\"#0 \\\/usr\\\/local\\\/www\\\/owncloud\\\/apps\\\/files_sharing\\\/lib\\\/sharedstorage.php(90): OC\\\\Files\\\\View->getPath(33084)\\n#1 \\\/usr\\\/local\\\/www\\\/owncloud\\\/apps\\\/files_sharing\\\/lib\\\/sharedstorage.php(581): OC\\\\Files\\\\Storage\\\\Shared->init()\\n#2 \\\/usr\\\/local\\\/www\\\/owncloud\\\/lib\\\/private\\\/files\\\/storage\\\/wrapper\\\/wrapper.php(381): OC\\\\Files\\\\Storage\\\\Shared->getCache('', Object(OCA\\\\Files_Trashbin\\\\Storage))\\n#3 \\\/usr\\\/local\\\/www\\\/owncloud\\\/lib\\\/private\\\/files\\\/view.php(1302): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Wrapper->getCache('')\\n#4 \\\/usr\\\/local\\\/www\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/connector\\\/sabre\\\/serverfactory.php(124): OC\\\\Files\\\\View->getFileInfo('')\\n#5 [internal function]: OCA\\\\DAV\\\\Connector\\\\Sabre\\\\ServerFactory->OCA\\\\DAV\\\\Connector\\\\Sabre\\\\{closure}(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#6 \\\/usr\\\/local\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Object(Closure), Array)\\n#7 \\\/usr\\\/local\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(446): Sabre\\\\Event\\\\EventEmitter->emit('beforeMethod', Array)\\n#8 \\\/usr\\\/local\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(248): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#9 \\\/usr\\\/local\\\/www\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(55): Sabre\\\\DAV\\\\Server->exec()\\n#10 \\\/usr\\\/local\\\/www\\\/owncloud\\\/remote.php(138): require_once('\\\/usr\\\/local\\\/www\\\/...')\\n#11 {main}\",\"File\":\"\\\/usr\\\/local\\\/www\\\/owncloud\\\/lib\\\/private\\\/files\\\/view.php\",\"Line\":1639}","level":3,"time":"2016-03-09T23:08:44+00:00","method":"PROPFIND","url":"\/remote.php\/webdav\/owncloud"}
```

File: apps/files_sharing/lib/sharedstorage.php (hunk 3 above)
Comment (icewind1991) [inline apps/files_sharing/lib/sharedstorage.php]: :disappointed: so much for easy code reuse

Comment (icewind1991) [inline apps/files_sharing/lib/sharedstorage.php]: moved FailedStorage/FailedCache to core

**Comment by icewind1991** [comment]
@PVince81 @DeepDiver1975 can this be merged?

**Comment by jospoortvliet** [comment]
It doesn't apply on my 9.0 installation:

```
patch -p1 --dry-run < 23039.patch 
checking file apps/files_external/lib/failedcache.php
checking file apps/files_sharing/lib/sharedstorage.php
checking file apps/files_external/lib/config/configadapter.php
checking file apps/files_external/lib/failedcache.php
Reversed (or previously applied) patch detected!  Assume -R? [n] n
Apply anyway? [n] n
Skipping patch.
1 out of 1 hunk ignored
checking file apps/files_external/lib/failedstorage.php
checking file apps/files_sharing/lib/sharedstorage.php
Hunk #1 FAILED at 31.
1 out of 1 hunk FAILED
checking file lib/private/files/cache/failedcache.php
checking file lib/private/files/storage/failedstorage.php
```

**Comment by jospoortvliet** [comment]
Fascinatingly, without the --dry-run, it applied without any problems.... So - a :+1: from me, obviously I can't speak about code quality or anything like that but it worked without problems.

**Comment by danimo** [comment]
@icewind @PVince81 Will this cure these warnings?

``
{"reqId":"g5uBYpFadSaPeNKIkyqg","remoteAddr":"192.168.39.100","app":"no app in context","message":"Exception: {\"Exception\":\"OCP\\\\Files\\\\NotFoundException\",\"Message\":\"File with id \\\"4216\\\" has not been found.\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/owncloud\\\/apps\\\/files_sharing\\\/lib\\\/sharedstorage.php(90): OC\\\\Files\\\\View->getPath(4216)\\n#1 \\\/var\\\/www\\\/owncloud\\\/apps\\\/files_sharing\\\/lib\\\/sharedstorage.php(581): OC\\\\Files\\\\Storage\\\\Shared->init()\\n#2 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/files\\\/storage\\\/wrapper\\\/wrapper.php(381): OC\\\\Files\\\\Storage\\\\Shared->getCache('', Object(OCA\\\\Files_Trashbin\\\\Storage))\\n#3 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/files\\\/view.php(1302): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Wrapper->getCache('')\\n#4 \\\/var\\\/www\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/connector\\\/sabre\\\/serverfactory.php(124): OC\\\\Files\\\\View->getFileInfo('\\\/Molkentin_Dani...')\\n#5 [internal function]: OCA\\\\DAV\\\\Connector\\\\Sabre\\\\ServerFactory->OCA\\\\DAV\\\\Connector\\\\Sabre\\\\{closure}(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#6 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Object(Closure), Array)\\n#7 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(446): Sabre\\\\Event\\\\EventEmitter->emit('beforeMethod', Array)\\n#8 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(248): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#9 \\\/var\\\/www\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(55): Sabre\\\\DAV\\\\Server->exec()\\n#10 \\\/var\\\/www\\\/owncloud\\\/remote.php(138): require_once('\\\/var\\\/www\\\/ownclo...')\\n#11 {main}\",\"File\":\"\\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/files\\\/view.php\",\"Line\":1639}","level":3,"time":"2016-03-14T13:40:46+00:00","method":"PROPFIND","url":"\/remote.php\/webdav\/"}
```

**Comment by PVince81** [comment]
@danimo no idea. Which file is 4216, does it exist ?

I suggest we merge this first and then investigate the core issue.

**Comment by PVince81** [comment]
:+1:

**Comment by PVince81** [comment]
@karlitschek backport to 9.0 ?

**Comment by rullzer** [comment]
:+1: 
I'd say a backport is a very good idea indeed!

**Comment by karlitschek** [comment]
looks safe! please backport :+1:

**Comment by icewind1991** [comment]
stable9: https://github.com/owncloud/core/pull/23252

**Comment by despens** [comment]
This patch breaks cron:

```
$ sudo -u www-data php cron.php 
PHP Fatal error:  Call to undefined method OC\Files\Cache\FailedCache::correctFolderSize() in /var/www/owncloud/lib/private/files/cache/scanner.php on line 457
```

**Comment by jospoortvliet** [comment]
@despens sure this is caused by this patch?

**Comment by despens** [comment]
What makes me think it does:
- it worked before I applied the patch and
- it concerns the FailedCache class which is subject to the patch.

**Comment by jospoortvliet** [comment]
@icewind1991 see above - could this have broken something?

**Comment by PVince81** [comment]
@despens can you raise a new ticket to make sure this doesn't get lost since this PR is already merged ? Thanks

**Pull request merged**